### PR TITLE
feature interactive job

### PR DESCRIPTION
**Pull request by AymenFJA**
This PR is based on #2573 (the next step is summit) provides the resource config file to run RP within an interactive job on multiple compute nodes. Example on TACC Frontera: 

```shell
srun --pty -A xxxx -N 2 -n 56 -t 01:00:00 -p development /bin/bash -l #  ask for 2 nodes interactively on Frontera
python 00_getting_started.py # run RP within these requested resources.
```

Note this config can be used to submit RP within `sbatch job`. Example from Frontera login node: 

1. Create an sbatch shell (rp_example.sh): 

```shell

#!/bin/bash
#----------------------------------------------------

#SBATCH -J 00_getting_started           # Job name
#SBATCH -o 00_getting_started.o%j       # Name of stdout output file
#SBATCH -e 00_getting_started.e%j       # Name of stderr error file
#SBATCH -p development     # Queue (partition) name
#SBATCH -N 2               # Total # of nodes
#SBATCH -n 56              # Total # of mpi tasks
#SBATCH -t 01:00:00        # Run time (hh:mm:ss)
#SBATCH -A XXXX            # Project/Allocation name (req'd if you have more than 1)

export RADICAL_LOG_LVL="DEBUG"
export RADICAL_PROFILE="TRUE"
export RADICAL_PILOT_DBURL=mongodb://user:password@ip/db_name

python 00_getting_started.py tacc.frontera_interactive

```

2. `sbatch rp_example.sh`

### CODE REVIEW

File: src/radical/pilot/configs/resource_tacc.json [hunk 1]
```diff
@@ -1,5 +1,42 @@
...
 {
+    "frontera_interactive": {
+        "description"                 : "Petascale computing system at the Texas Advanced Computing Center (TACC)",
+        "notes"                       : "Specify this resource in the pilot description and run RP within a compute node(s)",
+        "schemas"                     : ["local", "ssh"],
+        "ssh"                         :
+        {
+            "job_manager_endpoint"    : "ssh://localhost/",
+            "filesystem_endpoint"     : "sftp://localhost/"
+        },
```
Comment (mtitov) [inline src/radical/pilot/configs/resource_tacc.json:11]: could be removed (and a corresponding value from `schemas`), since we do local runs anyway

File: src/radical/pilot/configs/resource_tacc.json [hunk 2]
```diff
@@ -1,5 +1,42 @@
...
+        "agent_scheduler"             : "CONTINUOUS",
+        "agent_spawner"               : "POPEN",
+        "launch_methods"              : {
+                                         "order": ["SRUN"],
+                                         "SRUN" : {
+                                             "pre_exec_cached": []
+                                         }
+                                        },
+        "pre_bootstrap_1"             :["export RP_APP_TUNNEL_ADDR=144.76.72.175:27017",
+                                        "echo  $RP_APP_TUNNEL_ADDR"],
```
Comment (mtitov) [inline src/radical/pilot/configs/resource_tacc.json:30]: is this address valid?
but anyway, this could be removed as well, tunneling is not needed here

File: src/radical/pilot/configs/resource_tacc.json [hunk 3]
```diff
@@ -1,5 +1,42 @@
...
+        "resource_manager"            : "SLURM",
+        "cores_per_node"              : 56,
+        "gpus_per_node"               : 0,
+        "agent_config"                : "default",
+        "agent_scheduler"             : "CONTINUOUS",
+        "agent_spawner"               : "POPEN",
+        "launch_methods"              : {
+                                         "order": ["SRUN"],
+                                         "SRUN" : {
+                                             "pre_exec_cached": []
```
Comment (mtitov) [inline src/radical/pilot/configs/resource_tacc.json:26]: do we need to add corresponding modules load here? (as for `frontera`)

Comment (AymenFJA) [reply src/radical/pilot/configs/resource_tacc.json:26]: loading modules like the one we have in the `non-interactive` resource config ones lead to an error in MPI binding, as it overrides the preloaded module by the interactive job.

Comment (mtitov) [reply src/radical/pilot/configs/resource_tacc.json:26]: I would think yes and no - (1) I see that user could be completely responsible which modules should be loaded, since it is an interactive mode and user is able to control the environment; (2) RP is responsible to set a correct environment (load modules) to run a corresponding LM. I would follow the 2nd case, and extra modules, which are required for user, should go into task's `pre_exec`

Let's also see what's @andre-merzky 's opinion on this

Comment (AymenFJA) [reply src/radical/pilot/configs/resource_tacc.json:26]: Good points, thanks Mikhail, happy to accommodate any.

Comment (andre-merzky) [reply src/radical/pilot/configs/resource_tacc.json:26]: I don't even understand the problem :-P  Why would the environment of the client interact with the environment of the LM?  Or, in other words, why does it matter if the client lives on a login node or a compute node?  After the bootstrapper run, the agent env should be basically identical (assuming the bootstrapper loads the same modules)...

**Comment by andre-merzky** [comment]
@AymenFJA : additional to the points made by @mtitov , could you please also add resource configs for summit (both for jsrun and prrte), please?  That would useful in the context of radical-collaboration/hpc-workflows/issues/154 .

File: src/radical/pilot/configs/resource_tacc.json (hunk 1 above)
Comment (AymenFJA) [reply src/radical/pilot/configs/resource_tacc.json:11]: fixed

File: src/radical/pilot/configs/resource_tacc.json (hunk 2 above)
Comment (AymenFJA) [reply src/radical/pilot/configs/resource_tacc.json:30]: fixed

**Comment by andre-merzky** [comment]
> @AymenFJA : additional to the points made by @mtitov , could you please also add resource configs for summit (both for jsrun and prrte), please? That would useful in the context of [radical-collaboration/hpc-workflows/issues/154](https://github.com/radical-collaboration/hpc-workflows/issues/154) .

It's also fine to do that in a separate PR, no need to hold up this one...

File: src/radical/pilot/configs/resource_ornl.json [hunk 4]
```diff
@@ -181,6 +181,56 @@
...
+                                         "module load   libzmq/4.3.3",
+                                         "module load   python/3.7-anaconda3",
+                                         "ulimit -u 65536"
+                                        ],
+        "valid_roots"                 : ["$MEMBERWORK/"],
+        "default_remote_workdir"      : "$MEMBERWORK/${LSB_PROJECT_NAME,,}",
+        "python_dist"                 : "default",
+        "virtenv_dist"                : "default",
+        "virtenv_mode"                : "local",
+        "rp_version"                  : "local",
```
Comment (mtitov) [inline src/radical/pilot/configs/resource_ornl.json:225]: just keep only one parameter: `"virtenv_mode"                : "local"` (other parameters will be set by RP)

**Comment by mtitov** [comment]
@andre-merzky in case of Summit, we need to think how to determine the value of SMT. Currently RP will assume that SMT is equal `1`, but if user did set `smt4` and will try to use all hwthreads (168) per node, RP would think that more resources requested than provided (during the scheduling step?)
https://github.com/radical-cybertools/radical.pilot/blob/b6d3f04ac8332b14f7604e561ddf5bbabfc8d9b8/src/radical/pilot/pmgr/launching/base.py#L766-L770
Should Aymen make a test run with `smt4` set and couple tasks each with 64 threads?

**Comment by andre-merzky** [comment]
Did we check the pilot env to see if the SMT setting is (directly or indirectly) available somewhere?  If not, can we let the RM run a small test on a compute node to count the number of available cores maybe? (`jsrun -n 1 /bin/sh -c "cat /proc/cpuinfo  | grep -c 'core id'"`)?

**Comment by mtitov** [comment]
@andre-merzky looks like with some hacks it is possible to get that info:

> run a small test on a compute node to count the number of available cores maybe? (jsrun -n 1 /bin/sh -c "cat /proc/cpuinfo | grep -c 'core id'")?

Was able to get info with `jsrun -n 1 /bin/sh -c "cat /proc/cpuinfo | grep -c 'processor'"`, BUT it counts available cores plus blocked/system_reserved cores, e.g., for `smt4` it gives `176`, but for `smt2` it gives `92` (`(42 * 2) + (2 * 4)`)

Found a proposed solution, but still would need to do some parsing
```
# smt4
$ jsrun -n1 -bpacked:1 /bin/bash -c 'echo $OMP_PLACES'
{0:4}

# smt2
$ jsrun -n1 -bpacked:1 /bin/bash -c 'echo $OMP_PLACES'
{0:2}

# smt1
$ jsrun -n1 -bpacked:1 /bin/bash -c 'echo $OMP_PLACES'
{0}
```

Thus during the bootstrap we can execute the command above (+some parsing) and put that info into some environment variable (`RADICAL_SMT`? and keep the same env variable in SAGA instead of the [current one](https://github.com/radical-cybertools/radical.saga/blob/devel/src/radical/saga/adaptors/lsf/lsfjob.py#L255))

and then use that in PMGRLaunching as `smt = os.environ.get('RADICAL_SMT') or system_architecture.get('smt', 1)`

**Comment by andre-merzky** [comment]
> Found a proposed solution, but still would need to do some parsing

Nice!  I mean, it is somewhat clumsy what we try to do here - but I think that node inspection is something we will need to think about anyway (for example for heterogeneous nodes), and it's good to have some precedence to play with.

> but still would need to do some parsing

```sh
$ for X in "{0:4}" "{0:2}" "{0}"; do echo -n "$X : "; expr "$X" : '.*\(.\)}' | tr 0 1; done
{0:4} : 4
{0:2} : 2
{0} : 1
```


> UPDATE: I was wrong with speaking about "during the bootstrap" - SMT should be in sync between the client (PMGRLaunching) and the agent (ResourceManager) parts, and for interactive mode, env variable above (RADICAL_SMT) should be set ahead before client starts (?). And we want to have SMT-determing command within the config? Just thinking out loud..

Yeah, it won't help SAGA really, let's leave that as it is for now...

File: src/radical/pilot/pmgr/launching/base.py [hunk 5]
```diff
@@ -286,6 +286,13 @@ def work(self, pilots):
         for pilot in pilots:
             resource = pilot['description']['resource']
             schema   = pilot['description']['access_schema']
+            if schema == 'interactive' and 'ornl' in resource:
+                smt_cmd = r""
+                smt_cmd += r"expr $(jsrun -n1 -bpacked:1 /bin/bash -c 'echo $OMP_PLACES') :"
+                smt_cmd += r" '.*\(.\)}' | tr 0 1;"
+                smt = ru.sh_callout(smt_cmd, shell=True)[0].strip()
+                os.environ["RADICAL_SMT"] = smt
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/pmgr/launching/base.py:294]: Sorry for late response!  I would prefer if we do not need to rely on the user to provide the export.  `RADICAL_SMT` is great if it exists, but otherwise I agree we should either inspect the system for the setting (as we discussed in some place) or fall back to the configs.  A prominent log message (`logger.info('used SMT setting: %d' % smt`) would be useful.

File: src/radical/pilot/session.py [hunk 6]
```diff
@@ -714,6 +714,8 @@ def _get_resource_sandbox(self, pilot):
                     for k,v in pilot['description'].items():
                         if v is None:
                             v = ''
+                        if k == 'project' and '_' in v and 'ornl' in resource:
+                            v = v.split('_')[0]
```
Comment (andre-merzky) [inline src/radical/pilot/session.py:718]: What is the purpose of this change?

File: src/radical/pilot/pmgr/launching/base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mtitov) [reply src/radical/pilot/pmgr/launching/base.py:294]: Andre, but such code with `jsrun` command we shouldn't have in PMGRLaunching, right? It should get such info from the config (that what I would propose to have -> `cfg.pmgr_pre_launch`) and then execute it, thus it will set necessary env variable

File: src/radical/pilot/session.py (hunk 6 above)
Comment (mtitov) [reply src/radical/pilot/session.py:718]: this part was moved from outer scope, since it is related to PD attributes it is under `if '%' in sandbox_raw:` statement now, and it doesn't interfere with such case as having `_` in env variable name (`$MEMBERWORK/$LSB_PROJECT_NAME`)

File: src/radical/pilot/pmgr/launching/base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/pmgr/launching/base.py:294]: @mtitov and @andre-merzky discussed this again.  We decided that both on pmgr and on agent side we check `RADICAL_SMT`, and if not set fall back to the config value.  It is the user's responsibility to ensure either.

This also implies that `RADICAL_SMT` should be consistently set on the SAGA level, including in the SHELL adaptor.

[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/pmgr/launching/base.py]: @mtitov : I pushed a commit for the above, can you please check if that matches our discussion?
This should actually work w/o changes to the SAGA adaptors as `RADICAL_SMT` ends up in the pilot environment anyway and the bootstrapper picks it up from there.  But supporting in in the shell adaptor is cleaner and makes sense either way...

File: src/radical/pilot/pmgr/launching/base.py [hunk 7]
```diff
@@ -987,7 +980,7 @@ def _prepare_pilot(self, resource, rcfg, pilot, expand, tar_name):
         jd_dict.queue                 = queue
         jd_dict.candidate_hosts       = candidate_hosts
         jd_dict.file_transfer         = list()
-        jd_dict.environment           = dict()
+        jd_dict.environment           = {'RADICAL_SMT': smt}
```
Comment (mtitov) [inline src/radical/pilot/pmgr/launching/base.py:983]: can we move adding this env into line 991? - to keep env setup together

```python
jd_dict.environment['RADICAL_BASE'] = resource_sandbox
jd_dict.environment['RADICAL_SMT']  = smt
```

File: docs/source/faq.rst [hunk 8]
```diff
@@ -183,4 +183,40 @@ Number of concurrent RADICAL-Pilot scripts that can can be executed
...
+run an RP application in an interactive job allocation.  For example, the
+`ornl.summit` resource configuration defines an `interactive` access schema
+which can be used for that purpose.  A respective pilot description would look
+like this:
+
+    import radical.pilot as rp
+
+    pd = rp.PilotDescription()
+    pd.resource = 'ornl.summit'
+    pd.schema   = 'interactive'
```
Comment (mtitov) [inline docs/source/faq.rst:202]: With this PR we have a separate config for such case - `ornl.summit_interactive`.. there are two limitations to have it within `ornl.summit` - because of SAGA's Shell adaptor, which doesn't allow the following attributes ([link](https://github.com/radical-cybertools/radical.saga/blob/671dd9563228cad58580017963e573f83b91cf54/src/radical/saga/adaptors/shell/shell_job.py#L158-L175))
(1) use of `project` for `default_remote_workdir` (thus, for interactive mode it is `$MEMBERWORK/${LSB_PROJECT_NAME,,}`)
(2) `system_architecture`

Kinda, if we want to have a new schema in the current config for Summit, then these attributes could be added just to pass validation for SAGA Shell adaptor.

File: src/radical/pilot/pmgr/launching/base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mtitov) [reply src/radical/pilot/pmgr/launching/base.py]: added couple comments on that updates (mostly related to the provided example for Summit, and it is not the case for Frontera)

File: src/radical/pilot/configs/resource_ornl.json [hunk 9]
```diff
@@ -179,6 +179,53 @@
...
+                                         "module load   libzmq/4.3.3",
+                                         "module load   python/3.7-anaconda3"
+                                        ],
+        "pre_bootstrap_1"             : ["module load   gcc/9.1.0",
+                                         "module load   libzmq/4.3.3",
+                                         "module load   python/3.7-anaconda3",
+                                         "ulimit -u 65536"
+                                        ],
+        "valid_roots"                 : ["$MEMBERWORK/"],
+        "default_remote_workdir"      : "$MEMBERWORK/${LSB_PROJECT_NAME,,}",
```
Comment (andre-merzky) [inline src/radical/pilot/configs/resource_ornl.json:219]: I don't understand the `${NAME,,}` syntax - what am I missing about the commas?  The `ru.expand_env` uses a single `:` to delimit default values - I don't think that method is used here though...

File: src/radical/pilot/pmgr/launching/base.py (hunk 7 above)
Comment (andre-merzky) [reply src/radical/pilot/pmgr/launching/base.py:983]: Done.

File: docs/source/faq.rst (hunk 8 above)
Comment (andre-merzky) [reply docs/source/faq.rst:202]: Thanks,   I think this works for the time being.

File: src/radical/pilot/configs/resource_ornl.json (hunk 9 above)
Comment (mtitov) [reply src/radical/pilot/configs/resource_ornl.json:219]: that env variable contains a project name, but it is with capital letters, such format `${env_name,,}` gives a value in lower case (to be used for path)

File: docs/source/faq.rst (hunk 8 above)
Comment (mtitov) [reply docs/source/faq.rst:202]: should example be updates accordingly?
```
    pd.resource = 'ornl.summit_interactive'
    pd.schema   = 'interactive'
```
and/or
```
    pd.resource = 'tacc.frontera'
    pd.schema   = 'interactive'
```

File: examples/config.json [hunk 10]
```diff
@@ -110,6 +110,14 @@
         "cores"    : 56
     },
 
+    "tacc.frontera_interactive":{
```
Comment (mtitov) [inline examples/config.json:113]: @AymenFJA we don't have such config anymore

File: src/radical/pilot/pmgr/launching/base.py [hunk 11]
```diff
@@ -764,7 +771,7 @@ def _prepare_pilot(self, resource, rcfg, pilot, expand, tar_name):
                 cleanup = cleanup.replace('v', '')
 
         # estimate requested resources
-        smt = system_architecture.get('smt', 1)
+        smt = os.environ.get('RADICAL_SMT') or system_architecture.get('smt', 1)
```
[the remark this answers is not in the document]
Comment (mtitov) [reply src/radical/pilot/pmgr/launching/base.py:770]: @AymenFJA ping

File: docs/source/faq.rst (hunk 8 above)
Comment (mtitov) [reply docs/source/faq.rst]: updated description

**Review by andre-merzky — APPROVED** [review]
LGTM - thanks all!
